3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,301 @@
+        DigestUtils.md5Hex(b)
+      }.success
+      case "SHA-1" => { (b: Array[Byte]) =>
+        DigestUtils.sha1Hex(b)
+      }.success
+      case "SHA-256" => { (b: Array[Byte]) =>
+        DigestUtils.sha256Hex(b)
+      }.success
+      case "SHA-384" => { (b: Array[Byte]) =>
+        DigestUtils.sha384Hex(b)
+      }.success
+      case "SHA-512" => { (b: Array[Byte]) =>
+        DigestUtils.sha512Hex(b)
+      }.success
+      case fName => s"Unknown function $fName".failure
+    }
+
+  private def extractFields(piiFields: List[JObject]): Validation[String, List[PiiField]] =
+    piiFields.map {
+      case field: JObject =>
+        if (ScalazJson4sUtils.fieldExists(field, "pojo"))
+          extractString(field, "pojo", "field").flatMap(extractPiiScalarField)
+        else if (ScalazJson4sUtils.fieldExists(field, "json")) extractPiiJsonField(field \ "json")
+        else s"PII Configuration: pii field does not include 'pojo' nor 'json' fields. Got: [${compact(field)}]".failure
+      case json => s"PII Configuration: pii field does not contain an object. Got: [${compact(json)}]".failure
+    }.sequenceU
+
+  private def extractPiiScalarField(fieldName: String): Validation[String, PiiScalar] =
+    ScalarMutators
+      .get(fieldName)
+      .map(PiiScalar(_).success)
+      .getOrElse(s"The specified pojo field $fieldName is not supported".failure)
+
+  private def extractPiiJsonField(jsonField: JValue): Validation[String, PiiJson] = {
+    val schemaCriterion = extractString(jsonField, "schemaCriterion")
+      .flatMap(sc => SchemaCriterion.parse(sc).leftMap(_.getMessage))
+      .toValidationNel
+    val jsonPath = extractString(jsonField, "jsonPath").toValidationNel
+    val mutator = extractString(jsonField, "field")
+      .flatMap(getJsonMutator)
+      .toValidationNel
+    val validatedNel = (mutator |@| schemaCriterion |@| jsonPath)(PiiJson.apply)
+    validatedNel.leftMap(x => s"Unable to extract PII JSON: ${x.list.mkString(",")}")
+  }
+
+  private def getJsonMutator(fieldName: String): Validation[String, Mutator] =
+    JsonMutators
+      .get(fieldName)
+      .map(_.success)
+      .getOrElse(s"The specified json field ${fieldName} is not supported".failure)
+
+  private def extractString(jValue: JValue, field: String, tail: String*): Validation[String, String] =
+    ScalazJson4sUtils.extract[String](jValue, field, tail: _*).leftMap(_.getMessage)
+
+  private def extractStrategyFunction(config: JValue): Validation[String, String] =
+    ScalazJson4sUtils
+      .extract[String](config, "parameters", "strategy", "pseudonymize", "hashFunction")
+      .leftMap(_.getMessage)
+
+  private def matchesSchema(config: JValue, schemaKey: SchemaKey): Validation[String, JValue] =
+    if (supportedSchema.matches(schemaKey)) {
+      config.success
+    } else {
+      "Schema key %s is not supported. A '%s' enrichment must have schema '%s'."
+        .format(schemaKey, supportedSchema.name, supportedSchema)
```

Comment:
Quite strange that we use Scala interpolation everywhere and Java-ish `.format` in this single place.